display preview article cover when selected or dropped

johangiraud opened the pull request.

johangiraud committed:
display preview article cover when selected or dropped

johangiraud committed:
display preview article cover when selected or dropped

johangiraud committed:
display image when selected edit tag page

johangiraud committed:
display image when selected edit tag page

johangiraud committed:
center images article page in case they are too small

johangiraud committed:
center images article page in case they are too small

johangiraud committed:
use block and macro to call noise.js in required template files

johangiraud committed:
use block and macro to call noise.js in required template files

johangiraud committed:
prevent flash infos from being  visible when disappearing on mobile

johangiraud committed:
prevent flash infos from being  visible when disappearing on mobile

johangiraud committed:
fix parallax offset issue: backgroundPosition: _px => _%

johangiraud committed:
fix parallax offset issue: backgroundPosition: _px => _%

johangiraud committed:
add noise.js script to create draft

johangiraud committed:
add noise.js script to create draft

johangiraud committed:
tag page: display default class when tag cover is null

johangiraud committed:
tag page: display default class when tag cover is null

johangiraud committed:
fix meta description in article detail

johangiraud committed:
fix meta description in article detail

johangiraud committed:
change tag list text color

johangiraud committed:
change tag list text color

johangiraud committed:
fix create article image field height mobile devices

johangiraud committed:
fix create article image field height mobile devices

johangiraud committed:
move edit this article/draft button

johangiraud committed:
move edit this article/draft button

johangiraud committed:
duplicate edit button

johangiraud committed:
duplicate edit button

johangiraud committed:
move tag selector after the body text

johangiraud committed:
move tag selector after the body text

johangiraud committed:
add white background on .file label

johangiraud committed:
add white background on .file label

davidbgk commented on cafebabel/static/js/tag-detail.js:
```diff
@@ -2,34 +2,22 @@ const container = document.querySelector('.tag-header')
 
 function addBackgroundImage(container, imgUrl) {
-  if (imgUrl) {
+  if (imgUrl === 'None') {
```
You should not have to compare to `None` which is a python keyword, if you have that value within the template please submit a bug or turn it explicitly to an empty string to check the existence here instead of comparing to that string.

johangiraud replied:
That's exactly why I compared `imgUrl` with `None`. Cause I couldn't get it with `if (!value)` or something like that. How can I force `value=""`?

davidbgk replied:
In the template, adding `{{ article.property or '' }}`

johangiraud replied:
done.

johangiraud committed:
fix tag page if no image condition

johangiraud committed:
fix tag page if no image condition

davidbgk committed:
Add migration for old users, fix #191

The migration for 13k users is about 1 minute.

You have to retrieve the data locally first, see #190

vinyll committed:
a user can also login with a Django legacy password

johangiraud committed:
Merge branch 'master' into image-loading-display

davidbgk commented on cafebabel/templates/articles/tags/detail.html:
```diff
@@ -31,7 +31,7 @@
 {% block body_class %}tag-page{% endblock %}
 {% block body %}
-  <div class=tag-header data-image={{ tag.image_url }}></div>
+  <div class=tag-header data-image={{ tag.image_url or '' }}></div>
```
Can we switch to a regular `<img>` tag? Then [enhanced](https://en.wikipedia.org/wiki/Progressive_enhancement) by JavaScript?

johangiraud replied:
Why not, but I don't know how to do the parallax effect then. And I already spent too much time on this. Plus, after reading you wikipedia link I still don't really get what you mean by _Then enhanced by JavaScript_

vinyll replied:
I get what you were talking about now. Indeed having a `style="background-image: {{ tag.image_url }}"` could be one of those rare exception that could make some sense, and then apply the `parralax()` function onto it.

vinyll commented on cafebabel/static/css/styles.css:
```diff
@@ -845,41 +845,63 @@ button.follow.active:after {
 }
 
-.create-article-page canvas {
+@media screen and (min-width: 1200px) {
```
Are you sure about 1200px for adapting the behavior?

johangiraud replied:
yes definitely !

vinyll replied:
Ohh, because of the image width for article that is 1200px I suppose.
Can't we set a max-width/max-height in line 844 instead?
```css
.create-article-page .file canvas,
 .create-article-page .file {
   position: relative;
   width: 100%;
  max-width: 120rem;
  max-height: 46.7rem;
  …
```

vinyll replied:
> We dont want max-height to be 46.7rem on mobile :-(

I'm missing something here. I can't see the `max-height` that is currently applied for mobile and if there's none, I'm not aware of the side effect of a max-height when that height is not reached. Further details will be welcome to me :)

vinyll commented on cafebabel/static/css/styles.css:
```diff
@@ -845,41 +845,63 @@ button.follow.active:after {
 
-.create-article-page canvas {
+@media screen and (min-width: 1200px) {
+  .create-article-page canvas,
```
_canvas_ here is so generic we can't guess what it means unless digging exhaustively into the HTML.
It seems like `.file canvas` would suit – which would be a little more specific and explicit – but is it what you meant?

johangiraud replied:
ok, done. You are right

vinyll commented on cafebabel/static/js/image-preview.js:
```diff
@@ -0,0 +1,45 @@
+    const image = event.target.files[0]
+    const fileName = image.name
+    const reader = new FileReader()
+    const canvas = document.querySelector('canvas')
```
Is this the same `canvas` than the one where you have the noise effect on purpose or is it conflicting?

vinyll replied:
After re-reading the review I suppose you are trying to point the `canvas` that is inside the `.file`. In which case you could use a selector such as `document.querySelector('.file canvas')`

johangiraud replied:
right, need more consistency. Done.

vinyll commented on cafebabel/static/css/styles.css:
```diff
@@ -845,41 +845,63 @@ button.follow.active:after {
+  }
+}
+
+.file canvas {
```
I see you are using a very different selector than `.create-article-page canvas`. I therefore suppose this is not specific to _create-article-page_ and it should only apply to canvas within _.file_ while the other should apply to any canvas on the page.
🤔

johangiraud replied:
right. Fixed

johangiraud replied:
Oh no wait, here was the reason: I have now two different _noise_ effects: one on _create-article-page_ AND one on _tag-page_. That's why I have to precise the one for articles (cause I want a fixed width)

vinyll commented on cafebabel/static/css/styles.css:
```diff
@@ -1358,6 +1379,17 @@ h1.edit input {
+  max-width: 100%;
+}
+
 .article-page article h1 + p {
```
This is a nice self-explanatory selector BTW 👍

vinyll commented on cafebabel/static/js/image-preview.js:
```diff
@@ -0,0 +1,45 @@
+/* display picture file name after selection on profile edit https://tympanus.net/codrops/2015/09/15/styling-customizing-file-inputs-smart-way/ */
+const fileInput = document.querySelector('.file input[type="file"]')
+const outputImage = document.querySelector('.file output')
+const sizeInfo = fileInput.parentElement.querySelector('small')
```
Why not selecting `document.querySelector('.file small')` as you did for `outputImage`?

johangiraud replied:
sure.

vinyll commented on cafebabel/static/js/tag-detail.js:
```diff
@@ -2,34 +2,22 @@ const container = document.querySelector('.tag-header')
     const scrolledHeight = window.pageYOffset
-    const top = container.offsetTop - height * 2
-    container.style.backgroundPositionY = `${(top - scrolledHeight) / 3}px`
+    container.style.backgroundPositionY = 50 + 0.2 * scrolledHeight + '%'
```
template literal ftw: ```… = `${50 + 0.2 * scrolledHeight}%` ```

johangiraud replied:
ok done

vinyll commented on cafebabel/templates/_macro-extra-js.html:
```diff
@@ -0,0 +1,3 @@
+{% macro extra_js(file_path) %}
+  <script src={{ file_path }}></script>
+{% endmacro %}
```
newline missing Github says.
You should check for the option in you editor ;)

vinyll commented on cafebabel/templates/articles/detail.html:
```diff
@@ -14,13 +14,13 @@
   {{ meta('property=og:type', 'article') }}
-  {{ meta('property=og:description', 'pan-European news, features and opinions') }}
+  {{ meta('property=og:description', article.summary) }}
   {{ meta('property=og:site_name', 'Cafébabel') }}
```
This should be the value by default in the base template. Is it going to change per page?
If not it should just be hardcoded in the base.

johangiraud replied:
ha in `base.html`?

vinyll replied:
Yes, I suppose in _base.html_

vinyll commented on cafebabel/templates/articles/detail.html:
```diff
@@ -104,6 +104,7 @@ <h1>{{ article.title }}</h1>
       {% endfor %}
     </ul>
   {% endif %}
+  {{ self.edit_button() }}
```
Is the purpose of this to display twice the edit button?

johangiraud replied:
oui ! Made it on purpose

vinyll commented on cafebabel/templates/articles/drafts/create.html:
```diff
@@ -1,14 +1,20 @@
 
-{% block extra_js %}<script src={{ url_for('static', filename='js/articles-tags.js') }}></script>{% endblock %}
+{% block extra_js %}
+  {{ extra_js(url_for('static', filename='js/noise.js')) }}
```
Would it work if you were writing `{{ extra_js('js/noise.js') }}` here and deport the `url_for` within your macro?
If not there isn't much sense to write `{{ extra_js(` instead of `<script src=`

vinyll replied:
Another option would be to write `{{ extra_js(['js/noise.js', 'js/articles-tags.js', 'js/image-preview.js']) }}` and have a `for…endfor` in the macro.

johangiraud replied:
write `{{ extra_js('js/noise.js') }}` doesn't work :-(

vinyll replied:
The question may sounds silly, but did you plainly try `extra_js('js/noise.js')` with the current code or did you try to adapt the macro function with something like the code below?

```
{% macro extra_js(file_path) %}
  <script src="{{ url_for('static', filename=file_path) }}"></script>
{% endmacro %}
```

vinyll replied:
If it were not working, we could probably keep `{{ extra_js(url_for('static', filename='js/noise.js')) }}` in plain html:`<script src={{url_for('static', filename='js/noise.js')) }}></script>` as we don't gain anything; but I may be missing something here.

johangiraud replied:
doesn't work like that: `extra_js('js/noise.js')`. It would work with `extra_js('/static/js/noise.js')` though.
Hum you are right, we don't really gain anything here.

vinyll commented on cafebabel/templates/articles/drafts/create.html:
```diff
@@ -1,14 +1,20 @@
 {% extends "base.html" %}
 {% import "articles/macros.html" as macros without context %}
+{% from "_macro-extra-js.html" import extra_js without context %}
```
I think this `extra_js` could get into the _meta-macros.html file as it is an HTML meta.

However if you had a reason to think it different, we should challenge the filename. Wondering about what convention we are making with the macro names.
We have: 
- /macros.html
- /_forms-macros.html
- /_meta-macros.html

Maybe we could rename this to _extrajs-macros.html

johangiraud replied:
good point. I see what you mean when it comes to rename files for consistency. But I don't understand what you mean with: I think this `extra_js` could get into the _meta-macros.html file as it is an HTML meta. ?

johangiraud replied:
I renamed the file.

vinyll replied:
I meant that `extra_js()` function could also be in the _meta-macro.html file instead of being all alone in its file

johangiraud replied:
ah ok, well it's not a _meta_ per say

vinyll commented on cafebabel/templates/articles/drafts/create.html:
```diff
@@ -1,14 +1,20 @@
-    <label for=image>Click here to add a cover picture or Drag & drop your pic here <small>Recommended file size: less than 1Mo</small></label>
+    <canvas class=noise></canvas>
+    <output></output>
+    <label for=image>Click here to add a cover picture or Drag & drop your pic here <small>Recommended file size: 1200x467px, less than 1Mo</small></label>
```
Wondering how we should round this for the end user. 1200x500px?

johangiraud replied:
Previous website picture format was 1180x461px. That's why I would like to keep the proportions. What do you think about that @vinyll ?

vinyll replied:
For sure you're right that we should keep the proportions for the existing articles.
My question is more about the dimensions we ask the user to make for new articles; asking the user to make a picture 467px height is so precise that it sounds restrictive. Using a rounded value like 500px (or eventually 450px) could ease that feeling.
But it may be interesting to just quickly ask the editors around instead ;)

johangiraud replied:
Indeed, it would be easier for editors. But if we write a velue like 500px, then we will get images with a 500px height. Won't fit previous format then ? I don't get it.

vinyll commented on cafebabel/templates/articles/drafts/edit.html:
```diff
@@ -1,21 +1,30 @@
+  {{ extra_js(url_for('static', filename='js/noise.js')) }}
+  {{ extra_js(url_for('static', filename='js/articles-tags.js')) }}
+  {{ extra_js(url_for('static', filename='js/image-preview.js')) }}
+{% endblock %}
```
This is so similar to articles/edit.html that you may want to refactor and have this template inheriting from articles/edit.html.
Specific elements could be customized through _blocks_ like we did for draft/detail.html

vinyll commented on cafebabel/templates/articles/drafts/edit.html:
```diff
@@ -1,21 +1,30 @@
+      <canvas class=noise></canvas>
+      <output></output>
+    {% endif %}
+    <label for=image>Click here to add a cover picture or Drag & drop your pic here <small>Recommended file size: 1200x467px, less than 1Mo</small></label>
```
This copy-paste comforts the idea of refactoring

davidbgk committed:
Refactor the handling of uploads, fix #188

Note: a more consistent validation (not restricted to files) will be performed during #187

Side-effect: fixes #179 too!

vinyll commented on cafebabel/templates/articles/tags/edit.html:
```diff
@@ -1,18 +1,27 @@
+      <canvas class=noise></canvas>
+      <output></output>
     {% endif %}
+    <label for=image>Click here to add a cover picture or Drag & drop your pic here <small>Recommended file size: 1200x467px, less than 1Mo</small></label>
```
And again.
If for some reason you were not doing the refactoring, it should **at least** be a macro so that you don't forget one text when copy-pasting here and there.

vinyll commented on cafebabel/templates/users/edit.html:
```diff
@@ -1,7 +1,7 @@
 {% extends 'base.html' %}
 
-{% block extra_js %}<script src={{ url_for('static', filename='js/edit-profile.js') }}></script>{% endblock %}
+{% block extra_js %}<script src={{ url_for('static', filename='js/image-preview.js') }}></script>{% endblock %}
```
This extra_js block is not using the macro you had written.

johangiraud replied:
ah oui. my bad

johangiraud committed:
corrections after reviews 1/2

johangiraud committed:
corrections after reviews 1/2

johangiraud committed:
Merge branch 'image-loading-display' of github.com:cafebabel/cafebabel.com into image-loading-display

davidbgk committed:
Post-review fixes, refs #188

davidbgk committed:
Post-review fixes, refs #191

davidbgk committed:
Better django-related migrations naming, refs #191

johangiraud committed:
rename extra_js macro for more consistency

johangiraud committed:
rename extra_js macro for more consistency

johangiraud committed:
use macri extrajs in users/edit

johangiraud committed:
use macri extrajs in users/edit

johangiraud commented:
@vinyll about that:

> I get what you were talking about now. Indeed having a `style="background-image: {{ tag.image_url }}"` could be one of those rare exception that could make some sense, and then apply the `parralax()` function onto it.

I first made it this way: 
![capture d ecran 2018-01-31 a 10 17 46](https://user-images.githubusercontent.com/32518627/35614866-c4c7de48-0670-11e8-80f0-7fb2694e38e4.png)
(refering to this commit https://github.com/cafebabel/cafebabel.com/pull/152/commits/b12ea62efdfad6ccb83030aab8d960694a5044d3)
and you or @davidbgk said being strongly against that. That's why I changed it and passed it through js with `data-image`. What did change your mind ?

johangiraud committed:
add forgotten extrajs import

johangiraud committed:
add forgotten extrajs import

vinyll commented:
https://github.com/cafebabel/cafebabel.com/pull/178#issuecomment-361873001
> and you or @davidbgk said being strongly against that. That's why I changed it and passed it through js with data-image. What did change your mind ?
To make it clear, I'm not **strongly** against it ;)
These are different ways to reach a similar goal and I approve the way you are doing it now with `data-image` attribute. @davidbgk was challenging yesterday the idea that the JS may break or be available in which case the tag image would cease appearing. Possible answers are that it doesn't matter if the image doesn't display and it would still look good enough; other like David was suggesting could be to have the image that would displayed anyways and the parralax would be applied unless the JS would fail loading; or very probably other ideas too 🙂

vinyll commented on cafebabel/templates/users/edit.html:
```diff
@@ -1,7 +1,10 @@
-{% from "users/macros.html" import social_field, default_profile_picture with context %}
 {% extends 'base.html' %}
+{% from "users/macros.html" import social_field, default_profile_picture with context %}
+{% from "_extrajs-macros.html" import extra_js without context %}
```
BTW, I think writing `{% from "_extrajs-macros.html" import extra_js %}` should do the job and `without context` is optional.

johangiraud replied:
ok. Done.

johangiraud committed:
make upload field a macros

johangiraud committed:
make upload field a macros

johangiraud commented on cafebabel/templates/articles/macros.html:
```diff
@@ -34,3 +34,20 @@
   </div>
 {%- endmacro %}
+
+{%- macro image_upload(article) -%}
```
I didn't know how to get `article` dynamic. In tag/edit, I would have liked to use this macros while replacing `article` by `tag`. 
I tried something like:
`<output><img src={{ {{ article}}.image_url }} alt="Image for {{ {{ article}}.name }}">`
but it obviously doesn't work 👎 
any suggestion @vinyll ?

vinyll reviewed (commented):
The `.file canvas` things seem more consistent to me :+1:

Commit:
Merge branch 'master' into image-loading-display

johangiraud committed:
add default image meta

johangiraud committed:
add default image meta

johangiraud committed:
remove with context from import extrajs

johangiraud committed:
remove with context from import extrajs

johangiraud commented:
> Ohh, because of the image width for article that is 1200px I suppose.
Can't we set a max-width/max-height in line 844 instead?
```
.create-article-page .file canvas,
 .create-article-page .file {
   position: relative;
   width: 100%;
  max-width: 120rem;
  max-height: 46.7rem;
```
We dont want max-height to be 46.7rem on mobile :-(

johangiraud committed:
Merge branch 'image-loading-display' of github.com:cafebabel/cafebabel.com into image-loading-display

johangiraud committed:
fix homepage test

johangiraud committed:
fix homepage test

johangiraud committed:
refactor extra_js macros

johangiraud committed:
refactor extra_js macros

johangiraud committed:
fix article upload cover pic field style

johangiraud committed:
fix article upload cover pic field style

johangiraud committed:
Merge branch 'master' into image-loading-display

johangiraud committed:
rename canvas const to avoid confusing with the other canvas const

johangiraud committed:
change wording for file size wanted + use macros in tag/edit

johangiraud committed:
fix conflicts

johangiraud committed:
remove with context

johangiraud committed:
fix profile page image upload label

johangiraud committed:
add placeholder image to article list

johangiraud committed:
fix: profile-edit hide social network field when click elsewhere

johangiraud committed:
add square format recommandation for image profile

johangiraud committed:
fix test extra_js undefined

johangiraud committed:
inverse color existing translation <=> to-translate in article/detail

johangiraud committed:
switch follow button to social icons

johangiraud committed:
remove follow button styles

vinyll commented on cafebabel/templates/base.html:
```diff
@@ -56,15 +56,12 @@
-        <li><a href=#>Instagram</a></li>
-        <li><a href=#>LinkedIn</a></li>
-        <li><a href=#>Youtube</a></li>
+        <li><a href=# class=icon-facebook></a></li>
```
Still wondering what is the indent size on this project for HTML. Do you go for 2 or 4 spaces?

johangiraud replied:
2, but I am not going to re-indent manually all the files we have been working on before we set the indent to 2, am I? @vinyll

johangiraud replied:
Possible to add review so that I can merge? @vinyll @davidbgk

johangiraud committed:
fix indent in base.html

davidbgk commented on cafebabel/templates/base.html:
```diff
@@ -79,13 +76,12 @@
 
 <footer>
-  <p>© Cafébabel — {{ get_year() }}</p>
+  <p>© Cafébabel — 2017</p>
```
I think you failed a merge here.

johangiraud replied:
ah ok, I fixed it. But there were other chances regarding this chance, right ?

johangiraud replied:
changes*, change*

johangiraud committed:
add dynamic year footer

vinyll committed:
refactored drafts edit template into draft create template

davidbgk commented on cafebabel/templates/articles/macros.html:
```diff
@@ -37,7 +37,7 @@
 {%- macro image_upload(object) -%}
   <div class=file>
-    {% if object.image_filename %}
+    {% if object.has_image %}
```
It scares me we have something like that here, as for the `get_date` a merge has probably failed somewhere…

vinyll replied:
Mhhh… I manually typed `has_image` here. Do you mean it should remain `image_filename` in this branch?

davidbgk replied:
No, I mean we should not have a reference to `image_filename` in the template at first. Which commit added it? There is no mention on `master` AFAIK.

vinyll replied:
Well, Im afraid it's messed up then…
<img width="283" alt="screen shot 2018-02-02 at 2 35 43 pm" src="https://user-images.githubusercontent.com/145172/35753714-6bf09046-0826-11e8-8910-49a63371baed.png">
<img width="233" alt="screen shot 2018-02-02 at 2 35 27 pm" src="https://user-images.githubusercontent.com/145172/35753715-6c01ce60-0826-11e8-893b-32448dbc1edd.png">

vinyll committed:
refactored drafts edit template into article create template

Commit:
Inactive socials medias when input loses the focus

Commit:
Merge branch 'image-loading-display' of github.com:cafebabel/cafebabel.com into image-loading-display

johangiraud committed:
add <img> fallback then add background image with js in tag/detail

johangiraud committed:
fix social network icons position and size in profile/detail

davidbgk commented on cafebabel/static/js/tag-detail.js:
```diff
@@ -1,35 +1,20 @@
 const container = document.querySelector('.tag-header')
-const imgUrl = container.getAttribute('data-image')
+const imgUrl = container.querySelector('img')
```
It looks a bit too vague to me, why not `.tag-header img`?

johangiraud replied:
ok

davidbgk commented on cafebabel/static/js/tag-detail.js:
```diff
@@ -1,35 +1,20 @@
-
-  return img.height
+  if (!imgUrl) return
+  container.querySelector('img').classList.add('hidden')
```
Why not reuse `imgUrl`?

johangiraud replied:
ok. and I changed `imgUrl` by `bgImg` => more sense, right ?

davidbgk replied:
Even better: `backgroundImage`? :)

davidbgk commented on cafebabel/templates/articles/tags/detail.html:
```diff
@@ -31,7 +31,13 @@
+  {% else %}
+    <div class="tag-header default">
+  {% endif %}
+    </div>
```
From my experience with templating, it's easier on the long run to have if clauses encapsulating the whole `div` for small parts:

```
    {% if tag.has_image %}
      <div class="tag-header image">
        <img src={{ tag.image_url or '' }} alt="Cover picture for #{{ tag.name }}">
      </div>
   {% else %}
      <div class="tag-header default"></div>
   {% endif %}

```

johangiraud replied:
ok

davidbgk commented on cafebabel/templates/articles/tags/detail.html:
```diff
@@ -31,7 +31,13 @@
 {% block body %}
-  <div class=tag-header data-image={{ tag.image_url }}></div>
+  {% if tag.has_image %}
+    <div class="tag-header image">
```
What is the goal of `image` vs. `default`? Don't you want to apply the parallax to the default too?

johangiraud replied:
when `.image`: `.tag-page .tag-header.image + header{ margin-top:60vh}`
when `.default`: `.tag-page .tag-header.image + header{ margin-top:30vh}`

johangiraud committed:
minors corrections after review

johangiraud committed:
Merge branch 'master' into image-loading-display

johangiraud committed:
close form tag to fix draft/create page

johangiraud committed:
Merge branch 'image-loading-display' of github.com:cafebabel/cafebabel.com into image-loading-display
